src/workspacer.Sound/SoundManager.cs

```diff
@@ -0,0 +1,393 @@
+            if (_mmDeviceEnumerator == null)
+            {
+                throw new Exception("Creating device enumerator failed");
+            }
+
+            _mmDeviceEnumerator.RegisterEndpointNotificationCallback(this);
+        }
+
+        public SoundClient RegisterClientForDefault(DeviceType type, params DeviceRole[] roles)
+        {
+            var deviceId = GetDefaultDeviceId(type, roles);
+            var client = RegisterClient(deviceId, type, roles);
+
+            return client;
+        }
+
+        public SoundClient RegisterClientForDeviceId(string deviceId)
+        {
+            return RegisterClient(deviceId);
+        }
+
+        private SoundClient RegisterClient(string deviceId, DeviceType? type = null, params DeviceRole[] roles)
+        {
+            var client = CreateClient(deviceId, type, roles);
+
+            //Map events
+            client.VolumeChanged += HandleVolumeChanged;
+            client.MuteChanged += HandleMuteChanged;
+
+            _clients.Add(client);
+
+            return client;
+        }
+
+        private void HandleMuteChanged(DeviceInfo device, bool newState)
+        {
+            MuteChanged?.Invoke(device, newState);
+        }
+
+        private void HandleVolumeChanged(DeviceInfo device, float newState)
+        {
+            VolumeChanged?.Invoke(device, newState);
+        }
+
+        public IEnumerable<SoundClient> GetClientsForDeviceId(string deviceId)
+        {
+            return _clients.Where(x => x.DeviceInfo.Id == deviceId);
+        }
+
+        public string GetDefaultPlaybackDeviceId(DeviceRole role = DeviceRole.Multimedia)
+        {
+            return GetDefaultDeviceId(DeviceType.Render, role);
+        }
+
+        public string GetDefaultRecordingDeviceId(DeviceRole role = DeviceRole.Communications)
+        {
+            return GetDefaultDeviceId(DeviceType.Capture, role);
+        }
+
+        public string GetDefaultDeviceId(DeviceType type, params DeviceRole[] roles)
+        {
+            if (_mmDeviceEnumerator == null)
+            {
+                return null;
+            }
+
+            IMMDevice device = null;
+            foreach (var role in roles)
+            {
+                _mmDeviceEnumerator.GetDefaultAudioEndpoint(type.ConvertToEDataFlow(), role.ConvertToERole(), out device);
+                if(device != null)
+                {
+                    break;
+                }
+            }
+
+            var id = string.Empty;
+            device?.GetId(out id);
+            return id;
+        }
+
+        private SoundClient CreateClient(string deviceId, DeviceType? type = null, params DeviceRole[] roles)
+        {
+            if (_mmDeviceEnumerator == null)
+            {
+                return null;
+            }
+
+            _mmDeviceEnumerator.GetDevice(deviceId, out var device);
+            if (device == null)
+            {
+                return null;
```

Comment:
Is this expected to happen under normal operations? That is, for an incorrect device-id to be passed in? And if so, are we sure simply returning `null` is the most appropriate way to handle it?

Not saying we *must* change this, but I'm just curious about what sort of considerations was put into this use-case, since it was so clearly handled :smile: